### ISSUE TITLE
Not supported for different image lengths and widths

### ISSUE THREAD

**Issue by zhirui-gao** [ktertikas/part_nerf#11]
Thanks for your released code!
But I found it does not support for different image lengths and widths. The index will be out of bound (partnerf/dataset/camera.py):
```
ray_origins = ray_origins[idxs_rows, idxs_cols, ...]
ray_directions = ray_directions[idxs_rows, idxs_cols, ...]
```
        

The code can run to the end if I fix it directly like this:
```
ray_origins = ray_origins[idxs_cols, idxs_rows, ...]
ray_directions = ray_directions[idxs_cols, idxs_rows, ...]
```
But I don't know if it is correct.

**Comment by ktertikas** [comment]
Hi @zhirui-gao, thank you for the interest in our work!

Can you send me the config that you are trying to run and an example dataset? I don't think there is a bug in the camera part of the code, but I am happy to have a look at it :)

**Comment by zhirui-gao** [comment]
Thanks for your reply! Here is my training config.
```
train_data:
  dataset_factory: "RaysMasksColorsIndexedDataset"
  collection_type: "lego_dataset"
  dataset_directory: "./dataset"
  images_folder: "images"
  cameras_folder: "cameras"
  masks_folder: "masks"
  splits_file: ""
  image_size: [400, 300]
  near: 0.5
  far: 4.5
  rand: true
  n_rays: 1024
  n_samples: 64
  sampling_type: "equal"
  splits: ["train", "val"]
train:
  num_workers: 2
  batch_size: 2
val_data: null
validation: null
model:
  coarse_fine: false
  shape_embedding_network:
    type: "simple"
    embedding_size: 128
    num_embeddings: 1
    max_norm: 1.
  texture_embedding_network:
    type: "simple"
    embedding_size: 128
    num_embeddings: 1
    max_norm: 1.
  shape_decomposition_network:
    type: "simple"
    num_parts: 8
    embedding_size: 128
    output_size: 128
    encoder:
      type: "simple"
  texture_decomposition_network:
    type: "simple"
    num_parts: 8
    embedding_size: 128
    output_size: 128
    encoder:
      type: "simple"
  structure_network:
    layers:
      - "translations:embedding"
      - "rotations:embedding"
      - "scale:embedding"
    scale_max_a: 0.6
  occupancy_network:
    type: "masked_occ"
    embedding_size: 128
    num_blocks: 2
    output_dim: 257
    chunk_size: 50000
    sharpness_inside: 100
    sharpness_outside: 100
  color_network:
    type: "hard"
    encoder:
      type: "residual"
      proj_dims: [[510, 256, 256],[256, 256, 256],[256, 128, 64]]
      out_dims: 3
    dir_coord_system: "primitive"
    dir_proj_dims: 20
  ray_associator:
    type: "occupancy"
    implicit_threshold: 0.5
renderer:
  type: "occ_single_nerf"
  white_background: False
loss:
  type:
    - "mse_loss_positive"
    - "mask_loss"
    - "volume_consistency_loss"
    - "ray_max_point_crossentropy"
    - "ray_max_point_crossentropy_coarse"
    - "ray_coverage_loss"
    - "ray_overlapping_loss_coarse"
    - "shape_embedding_normalization_loss"
    - "texture_embedding_normalization_loss"
  weights:
    - 1.0
    - 1.0
    - 1.0
    - 1.0
    - 0.0001
    - 0.01
    - 0.01
    - 0.0001
    - 0.0001
  num_inside_rays: 32
scheduler:
  type: warmup_cosine
  warmup_steps: 300
  max_steps: 250000
  start_lr: 0.0
  eta_min: 0.000005
optimizer:
  lr: 0.0005
metric:
  type:
    - "psnr"
    - "scale"
    - "associator"
trainer:
  start_epoch: 0
  num_epochs: 250
  steps_per_epoch: 200
  grad_accumulation_steps: 1
  statslogger: ""
  project_name: "nerf_editor"
  experiment_name: null
  output_directory: "../outputs"
  save_checkpoint_every: 50
  run_validation_every: null
  train_visualize_every: 20
  seed: 23
```

**Comment by zhirui-gao** [comment]
When I set the image to have the same length and width, everything runs very well

**Comment by ktertikas** [comment]
Can you please also send me a single image / mask / cameras example to test?

**Comment by zhirui-gao** [comment]
> Can you please also send me a single image / mask / cameras example to test?

Thanks for your kind. This is the [link](https://drive.google.com/file/d/1zuSvIJP42cmAwc6XIH1tLBhMGC6hKZWA/view?usp=drive_link) of data.
Looking forward your feedback!

**Comment by zhirui-gao** [comment]
Hi, I change the line
https://github.com/ktertikas/part_nerf/blob/87dc4501902b8192a3be1664104ce560084627b5/src/part_nerf/dataset/camera.py#L139-L142
to:
```
v, u = torch.meshgrid(
            torch.linspace(0, self._original_H - 1, H),
            torch.linspace(0, self._original_W - 1, W),
            indexing='ij'
        )
```
I think it works for images with different widths and heights!

**Comment by ktertikas** [comment]
Hi @zhirui-gao, there is probably a small bug in the resizing of the images and masks (https://github.com/ktertikas/part_nerf/blob/87dc4501902b8192a3be1664104ce560084627b5/src/part_nerf/dataset/base.py#L110)  (will push a quick fix once the discussion is done, but it's just a matter of inverting the `img_size` and `mask_size` in the relevant places).
Regarding the meshgrid operation, I am pretty sure PyTorch has changed the default indexing in newer versions, are you sure that you are using the same version defined in the environment file (pytorch 1.8)? If you are using the newer pytorch version, then your suggested change will indeed fix your issues.

**Comment by zhirui-gao** [comment]
Thanks for your explanation. It has fixed my confusion and I will close the issues.
